tiny_dnn/util/serialization_functions.h

```diff
@@ -10,6 +10,129 @@
 #include <cereal/access.hpp>  // For LoadAndConstruct
 #include "tiny_dnn/tiny_dnn.h"
 
+namespace {
+
+/**
+ * size of layer, model, data etc.
+ * change to smaller type if memory footprint is severe
+ **/
+typedef std::uint32_t serial_size_t;
+
+typedef tiny_dnn::index3d<serial_size_t> shape3d_serial;
+
+template <class T>
+static inline cereal::NameValuePair<T> make_nvp(const char *name, T &&value) {
+  return cereal::make_nvp(name, value);
+}
+
+template <class Archive, typename T>
+void arc(Archive &ar, cereal::NameValuePair<T> &&arg) {
+  ar(arg);
+}
+
+template <class Archive,
+          typename std::enable_if<std::is_base_of<cereal::BinaryOutputArchive,
+                                                  Archive>::value>::type = 0>
+void arc(Archive &ar, cereal::NameValuePair<size_t> &&arg) {
+  cereal::NameValuePair<serial_size_t> arg2(arg.name, arg.value);
+  ar(arg2);
+}
+
+template <class Archive,
+          typename std::enable_if<std::is_base_of<cereal::BinaryInputArchive,
+                                                  Archive>::value>::type = 0>
+void arc(Archive &ar, cereal::NameValuePair<size_t> &&arg) {
+  cereal::NameValuePair<serial_size_t> arg2(arg.name, 0);
+  ar(arg2);
+  arg.value = arg2.value;
+}
+
+static serial_size_t to_serial(const size_t &src) {
+  return static_cast<serial_size_t>(src);
+}
+
+static size_t to_size(const serial_size_t &src) {
+  return static_cast<size_t>(src);
+}
+
+static inline shape3d_serial to_serial_shape(const tiny_dnn::shape3d &src) {
+  shape3d_serial dst;
+  dst.width_  = to_serial(src.width_);
+  dst.height_ = to_serial(src.height_);
+  dst.depth_  = to_serial(src.depth_);
+  return dst;
+}
+
+static inline tiny_dnn::shape3d to_size_shape(const shape3d_serial &src) {
+  tiny_dnn::shape3d dst;
+  dst.width_  = to_size(src.width_);
+  dst.height_ = to_size(src.height_);
+  dst.depth_  = to_size(src.depth_);
+  return dst;
+}
+
+template <class Archive,
+          typename std::enable_if<std::is_base_of<cereal::BinaryOutputArchive,
+                                                  Archive>::value>::type = 0>
+void arc(Archive &ar, cereal::NameValuePair<tiny_dnn::shape3d> &&arg) {
+  cereal::NameValuePair<shape3d_serial> arg2(arg.name,
+                                             to_serial_shape(arg.value));
+  ar(arg2);
+}
+
+template <class Archive,
+          typename std::enable_if<std::is_base_of<cereal::BinaryInputArchive,
+                                                  Archive>::value>::type = 0>
+void arc(Archive &ar, cereal::NameValuePair<tiny_dnn::shape3d> &&arg) {
```

Comment:
Is this function OK in this way? It seems to me `arg.value` isn't set after calling an `ar` function object.